here-naksha-lib-psql/src/commonMain/kotlin/naksha/psql/executors/query/WhereClauseBuilder.kt

```diff
@@ -160,6 +163,81 @@ class WhereClauseBuilder(private val request: ReadFeatures) {
         }
     }
 
+    private fun whereTags() {
+        val tagQuery = request.query.tags
+        if(tagQuery != null){
+            if (where.isNotEmpty()) {
+                where.append(" AND (")
+            } else {
+                where.append(" (")
+            }
+            whereNestedTags(tagQuery)
+            where.append(")")
+        }
+    }
+
+    private fun whereNestedTags(tagQuery: ITagQuery) {
+        when (tagQuery) {
+            is TagNot -> not(tagQuery.query, this::whereNestedTags)
+            is TagOr -> or(tagQuery.filterNotNull(), this::whereNestedTags)
+            is TagAnd -> and(tagQuery.filterNotNull(), this::whereNestedTags)
+            is TagQuery -> resolveSingleTagQuery(tagQuery)
+        }
+    }
+
+    private fun resolveSingleTagQuery(tagQuery: TagQuery) {
+        when (tagQuery) {
+            is TagExists -> {
+                where.append("$tagsAsJsonb ?? '${tagQuery.name}'")
+            }
+
+            is TagValueIsNull -> {
+                where.append("${tagValue(tagQuery)} = null")
+            }
+
+            is TagValueIsBool -> {
+                if (tagQuery.value) {
+                    where.append(tagValue(tagQuery, PgType.BOOLEAN))
+                } else {
+                    where.append("not(${tagValue(tagQuery, PgType.BOOLEAN)}})")
+                }
+            }
+
+            is TagValueIsDouble -> {
+                val valuePlaceholder = placeholderForArg(tagQuery.value, PgType.DOUBLE)
+                val doubleOp = resolveDoubleOp(
+                    tagQuery.op,
+                    tagValue(tagQuery, PgType.DOUBLE),
+                    valuePlaceholder
+                )
+                where.append(doubleOp)
+            }
+
+            is TagValueIsString -> {
+                val valuePlaceholder = placeholderForArg(tagQuery.value, PgType.STRING)
+                val stringEquals = resolveStringOp(
+                    StringOp.EQUALS,
+                    tagValue(tagQuery, PgType.STRING),
+                    valuePlaceholder
+                )
+                where.append(stringEquals)
+            }
+
+            is TagValueMatches -> {
+                val regex = tagQuery.regex
+                where.append("$tagsAsJsonb @?? '\$.${tagQuery.name} ? (@ like_regex \"${regex}\")'")
```

Comment:
Again, this is prone to SQL and ASCII zero injections, because not correctly escaped (remember, tags that start with `@` allow all unicode characters! For regular expression, we should review carefully, if there is no way of code execution or query manipulation.